lib/babe/epoch_handler.go

```diff
@@ -0,0 +1,155 @@
+
+	handleSlot handleSlotFunc
+}
+
+func newEpochHandler(epochNumber, firstSlot uint64, epochData *epochData, constants *constants,
+	handleSlot handleSlotFunc, keypair *sr25519.Keypair) (*epochHandler, error) {
+	// determine which slots we'll be authoring in by pre-calculating VRF output
+	slotToProof := make(map[uint64]*VrfOutputAndProof)
+	for i := firstSlot; i < firstSlot+constants.epochLength; i++ {
+		proof, err := claimPrimarySlot(
+			epochData.randomness,
+			i,
+			epochNumber,
+			epochData.threshold,
+			keypair,
+		)
+		if errors.Is(err, errOverPrimarySlotThreshold) {
+			continue
+		} else if err != nil {
+			return nil, fmt.Errorf("error running slot lottery at slot %d: %w", i, err)
+		}
+
+		slotToProof[i] = proof
+		logger.Debugf("epoch %d: claimed slot %d", epochNumber, i)
+	}
+
+	return &epochHandler{
+		epochNumber: epochNumber,
+		firstSlot:   firstSlot,
+		constants:   constants,
+		epochData:   epochData,
+		slotToProof: slotToProof,
+		handleSlot:  handleSlot,
+	}, nil
+}
+
+func (h *epochHandler) run(ctx context.Context, errCh chan<- error) {
+	currSlot := getCurrentSlot(h.constants.slotDuration)
+
+	// if currSlot < h.firstSlot, it means we're at genesis and waiting for the first slot to arrive.
+	// we have to check it here to prevent int overflow.
+	if currSlot >= h.firstSlot && currSlot-h.firstSlot > h.constants.epochLength {
+		logger.Warnf("attempted to start epoch that has passed: current slot=%d, start slot of epoch=%d",
+			currSlot, h.firstSlot,
+		)
+		errCh <- errEpochPast
+		return
+	}
+
+	// for each slot we're handling, create a timer that will fire when it starts
+	// we create timers only for slots where we're authoring
+	authoringSlots := getAuthoringSlots(h.slotToProof)
+
+	type slotWithTimer struct {
+		timer   <-chan time.Time
+		slotNum uint64
+	}
+
+	slotTimeTimers := []*slotWithTimer{}
+	for _, authoringSlot := range authoringSlots {
+		// ignore slots already passed
+		if authoringSlot < currSlot {
+			continue
+		}
+
+		startTime := getSlotStartTime(authoringSlot, h.constants.slotDuration)
+		slotTimeTimers = append(slotTimeTimers, &slotWithTimer{
+			timer:   time.After(time.Until(startTime)),
+			slotNum: authoringSlot,
+		})
+		logger.Debugf("start time of slot %d: %v", authoringSlot, startTime)
+	}
+
+	logger.Debugf("authoring in %d slots in epoch %d", len(slotTimeTimers), h.epochNumber)
+
+	for _, swt := range slotTimeTimers {
+		logger.Debugf("waiting for next authoring slot %d", swt.slotNum)
+
+		select {
+		case <-ctx.Done():
+			return
+		case <-swt.timer:
+			if _, has := h.slotToProof[swt.slotNum]; !has {
+				// this should never happen
+				logger.Errorf("no VRF proof for authoring slot! slot=%d", swt.slotNum)
+				continue
+			}
+
+			if err := h.handleSlot(h.epochNumber, swt.slotNum, h.epochData.authorityIndex, h.slotToProof[swt.slotNum]); err != nil { //nolint:lll
+				logger.Warnf("failed to handle slot %d: %s", swt.slotNum, err)
+				continue
+			}
```

Comment:
nit don't use the short err check form with a big line like this.

```suggestion
			err := h.handleSlot(h.epochNumber, swt.slotNum, 
				h.epochData.authorityIndex, h.slotToProof[swt.slotNum])
			if err != nil {
				logger.Warnf("failed to handle slot %d: %s", swt.slotNum, err)
				continue
			}
```